crates/ra_ide_api/src/imp.rs

```diff
@@ -243,21 +247,74 @@ impl db::RootDatabase {
-        let res = self
-            .find_all_refs(position)?
-            .iter()
-            .map(|(file_id, text_range)| SourceFileEdit {
-                file_id: *file_id,
+    ) -> Cancelable<Option<SourceChange>> {
+        let mut source_file_edits = Vec::new();
+        let mut file_system_edits = Vec::new();
+
+        let source_file = self.source_file(position.file_id);
+        let syntax = source_file.syntax();
+        // We are rename a mod
+        if let (Some(ast_module), Some(name)) = (
+            find_node_at_offset::<ast::Module>(syntax, position.offset),
+            find_node_at_offset::<ast::Name>(syntax, position.offset),
+        ) {
+            if let Some(module) = module_from_declaration(self, position.file_id, &ast_module)? {
+                let (file_id, module_source) = module.definition_source(self)?;
+                match module_source {
+                    ModuleSource::SourceFile(..) => {
+                        let move_file = FileSystemEdit::MoveFile {
+                            src: file_id,
+                            dst_source_root: self.file_source_root(position.file_id),
+                            dst_path: self
+                                .file_relative_path(file_id)
+                                .with_file_name(new_name)
+                                .with_extension("rs"),
+                        };
+                        file_system_edits.push(move_file);
+                    }
+                    ModuleSource::Module(..) => {}
+                }
+            }
+
+            let edit = SourceFileEdit {
+                file_id: position.file_id,
                 edit: {
                     let mut builder = ra_text_edit::TextEditBuilder::default();
-                    builder.replace(*text_range, new_name.into());
+                    builder.replace(name.syntax().range(), new_name.into());
                     builder.finish()
                 },
-            })
-            .collect::<Vec<_>>();
-        Ok(res)
+            };
+            source_file_edits.push(edit);
+        }
+        // rename references
+        else {
+            let edit = self
+                .find_all_refs(position)?
+                .iter()
+                .map(|(file_id, text_range)| SourceFileEdit {
+                    file_id: *file_id,
+                    edit: {
+                        let mut builder = ra_text_edit::TextEditBuilder::default();
+                        builder.replace(*text_range, new_name.into());
+                        builder.finish()
+                    },
+                })
+                .collect::<Vec<_>>();
+            if edit.is_empty() {
+                return Ok(None);
+            }
+
+            source_file_edits = edit;
+        }
+
+        return Ok(Some(SourceChange {
+            label: "rename".to_string(),
+            source_file_edits,
+            file_system_edits,
+            cursor_position: None,
+        }));
     }
```

Comment:
This is quite a bit of code now... 

Let's move it to the new `rename` module, and introduce two private helper functions there, `rename_mod` and `rename_reference`. See the `runables.rs` module for how such organization might look like.